### DISCUSSION TITLE
Endpoint with Verb (::verb) and parameter

### DISCUSSION THREAD

**Discussion by fredrikj31** [fastify/fastify#4324] [Q&A] [answered]
Hi.

I was wondering if it's possible to have an endpoint like this:
- `localhost:3000/v1/cars/carId:start`
- `localhost:3000/v1/cars/carId:stop`

Where in Fastify I think it would be structured like this:
- `localhost:3000/v1/cars/:carId::start`
- `localhost:3000/v1/cars/:carId::stop`

I can't figure out a way to structure the endpoint url. Is it possible to have an endpoint with Fastify which supports this kind og behavior?

Best regards Fredrik Johansen

**Comment by mcollina** [comment]
@ivan-tymoshenko do you remember if we support escaping the `:`? In case we should document it.

**Comment by ivan-tymoshenko** [comment]
@fredrikj31 Can you give me an example of an incoming url, please?

**Reply by fredrikj31** [reply]
An example of a incoming URL could be:
POST: `https://localhost:3000/v1/cars/d0c4c452-2da5-48fc-a410-98871e2cbc1c:start`

I want it to be able to both support the uuid and action (:start) Is this possible?

**Comment by ivan-tymoshenko** [comment, marked as answer]
@fredrikj31  If you just want a path containing a colon without declaring a parameter, use a double colon. For example, /name::customVerb will be interpreted as /name:customVerb.

But your case is different. As I understand, you want to have a parameter + static suffix. [find-my-way](https://github.com/delvedor/find-my-way), which is Fastify router, supports . and - chars as markers that the param name is ended. So you can natively define routes like that:
`localhost:3000/v1/cars/:carId-start`
`localhost:3000/v1/cars/:carId.stop`

If you don't want to use '.' or '-' in your path, you should use the regexp route.
`localhost:3000/v1/cars/:carId(.*)::start`
`localhost:3000/v1/cars/:carId(.*)::stop`
depends on your request url format.

I checked if it works, and it doesn't. There is a bug. It doesn't escape the double colon in a multi-parametric node.

**Reply by ivan-tymoshenko** [reply]
Fixed it in https://github.com/delvedor/find-my-way/pull/307.
@fredrikj31 You should wait for the next Fastify release and use these routes
`localhost:3000/v1/cars/:carId(.*)::start`
`localhost:3000/v1/cars/:carId(.*)::stop`

**Reply by fredrikj31** [reply]
Thanks for the good explanation, and in depth knowledge of how it all works. So the thing you are saying, is that it is not possible to make it like that, because it caused by a bug?

**Reply by fredrikj31** [reply]
> Fixed it in https://github.com/delvedor/find-my-way/pull/307.
> @fredrikj31 You should wait for the next Fastify release and use these routes
> `localhost:3000/v1/cars/:carId(.*)::start`
> `localhost:3000/v1/cars/:carId(.*)::stop`

Is there a set date on when this fastify version will come out? If so, do you know when? @ivan-tymoshenko

**Reply by ivan-tymoshenko** [reply]
> Thanks for the good explanation, and in depth knowledge of how it all works. So the thing you are saying, is that it is not possible to make it like that, because it caused by a bug?

Colon is a special char that marks start of param name. So if you want to use it as a regular char in path you should escape it with another colon. Escaping behavior is not working right now in your case, but it will work after we land the fix.

**Reply by ivan-tymoshenko** [reply]
> > Fixed it in https://github.com/delvedor/find-my-way/pull/307.
> > @fredrikj31 You should wait for the next Fastify release and use these routes
> > `localhost:3000/v1/cars/:carId(.*)::start`
> > `localhost:3000/v1/cars/:carId(.*)::stop`
> 
> Is there a set date on when this fastify version will come out? If so, do you know when? @ivan-tymoshenko 

You should better ask @mcollina, but I would expect the next release in the middle of October.

**Reply by ivan-tymoshenko** [reply]
Meanwhile as a workaround you can create routes like that
`localhost:3000/v1/cars/:carId(.*)start`
`localhost:3000/v1/cars/:carId(.*)stop`

and check/trim the `:` manually in route handler.

**Reply by fredrikj31** [reply]
Thanks for the workaround, with the short regex pattern. It worked by just trimming out the colon in the handler. Thank you so much, and also for the fast support and answers 👍

**Comment by jsumners** [comment]
I think you'll find using `:` in a URI path to be problematic https://www.rfc-editor.org/rfc/rfc3986#section-2.2

**Reply by ivan-tymoshenko** [reply]
We support params encoding/decoding for this case.
